lib/membership/membership.go

```diff
@@ -0,0 +1,60 @@
+/*
+Copyright 2020 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package membership provides interface for cluster membership management.
+package membership
+
+import (
+	"net"
+
+	"github.com/gravitational/satellite/lib/client"
+
+	"golang.org/x/net/context"
+)
+
+// ClusterMembership is interface to interact with a cluster membership service.
+type ClusterMembership interface {
+	// Members returns the list of cluster members.
+	Members() ([]ClusterMember, error)
+	// FindMember finds the member with the specified name.
+	FindMember(name string) (ClusterMember, error)
+	// Close closes the client.
+	Close() error
+	// Join attempts to join an existing cluster identified by peers.
+	// Replay controls if previous user events are replayed once this node has joined the cluster.
+	// Returns the number of nodes joined.
+	Join(peers []string, replay bool) (int, error)
+	// UpdateTags will modify the tags on a running member.
+	UpdateTags(tags map[string]string, delTags []string) error
+	// GetCoordinate returns the Serf Coordinate for a specific node
+	// GetCoordinate(node string) (*coordinate.Coordinate, error)
+}
+
+// ClusterMember is interface to interact with a cluster member.
+type ClusterMember interface {
+	// Dial attempts to create client connect to member.
+	Dial(ctx context.Context, caFile, certFile, keyFile string) (client.Client, error)
```

Comment:
Maybe it's just me but I find it a bit weird that "Dial" is a part of the cluster member interface. This means that when the caller deals with it, ClusterMember instance can be in the "non initialized" state i.e. require to call Dial first. Does this have to be a part of the interface? I would consider moving this into a separate public method that returns a ClusterMember instance.

Reply:
> This means that when the caller deals with it, ClusterMember instance can be in the "non initialized" state i.e. require to call Dial first.

Can you clarify what problems can occur when dealing with a non initialized member?

Would it make sense to have a `Dialer/RPCDialer` interface but still have `SerfMember` implement `Dialer`?